core/include/initcall.h

```diff
@@ -30,12 +30,67 @@ struct initcall {
 		{ .func = (fn), }
 #endif
 
+#define preinitcall_begin \
+			SCATTERED_ARRAY_BEGIN(preinitcall, struct initcall)
+#define preinitcall_end SCATTERED_ARRAY_END(preinitcall, struct initcall)
+
 #define initcall_begin	SCATTERED_ARRAY_BEGIN(initcall, struct initcall)
 #define initcall_end	SCATTERED_ARRAY_END(initcall, struct initcall)
 
 #define finalcall_begin	SCATTERED_ARRAY_BEGIN(finalcall, struct initcall)
 #define finalcall_end	SCATTERED_ARRAY_END(finalcall, struct initcall)
 
+/*
+ * The preinit_*(), *_init() and boot_final() macros are used to register
+ * callback functions to be called at different stages during
+ * initialization.
+ *
+ * Functions registered with preinit_*() are always called before functions
+ * registered with *_init().
+ *
+ * Functions registered with boot_final() are called before exiting to
+ * normal world the first time.
+ *
+ * Without virtualization this happens in the order of the defines below.
+ *
+ * However, with virtualization things are a bit different. boot_final()
+ * functions are called first before exiting to normal world the first
+ * time. Functions registered with boot_final() can only operate on the
+ * nexus. preinit_*() functions are called early before the first yielding
+ * call into the partition, in the newly created partition. *_init()
+ * functions are called at the first yielding call.
+ *
+ *  +-------------------------------+-----------------------------------+
+ *  | Without virtualization        | With virtualization               |
+ *  +-------------------------------+-----------------------------------+
+ *  | At the end of boot_init_primary_late() just before the print:     |
+ *  | "Primary CPU switching to normal world boot"                      |
+ *  +-------------------------------+-----------------------------------+
+ *  | 1. call_preinitcalls()        | In the nexus                      |
+ *  | 2. call_initcalls()           +-----------------------------------+
+ *  | 3. call_finalcalls()          | 1. call_finalcalls()              |
```

Comment:
`bootfinal` callbacks are called before `*_init` and `preinit` levels. It sounds strange to me

Looking at the registered `boot_final` callbacks, i think they are misused.
`
core/arch/arm/kernel/secure_partition.c:boot_final(sp_init_all);
core/arch/arm/kernel/boot.c:boot_final(release_external_dt);
core/drivers/crypto/se050/glue/i2c.c:boot_final(load_trampoline);
`
The `secure_partition.c` callbacks could be `early_init` level, there are no dependencies.
The `se050/i2c.c` callback should be called only once a VM has been created, no? (`driver_init_late` level)
As for `release_external_dt()`, maybe it should be simply explicitly called from `boot_init_primary_late()` instead of `call_finalcalls()`.